src/app/clusters/energy-preference-server/energy-preference-server.cpp

```diff
@@ -0,0 +1,221 @@
+                {
+                    return CHIP_NO_ERROR;
+                }
+                return err;
+            });
+        }
+        return CHIP_ERROR_INCORRECT_STATE;
+    case LowPowerModeSensitivities::Id:
+        if (lowPowerSupported == false)
+        {
+            return aEncoder.EncodeNull();
+        }
+
+        if (gsDelegate != nullptr)
+        {
+            return aEncoder.EncodeList([endpoint](const auto & encoder) -> CHIP_ERROR {
+                BalanceStruct::Type balance;
+                size_t index   = 0;
+                CHIP_ERROR err = CHIP_NO_ERROR;
+                while ((err = gsDelegate->GetLowPowerModeSensitivityAtIndex(endpoint, index, balance)) == CHIP_NO_ERROR)
+                {
+                    ReturnErrorOnFailure(encoder.Encode(balance));
+                    index++;
+                }
+                if (err == CHIP_ERROR_NOT_FOUND)
+                {
+                    return CHIP_NO_ERROR;
+                }
+                return err;
+            });
+        }
+        return CHIP_ERROR_INCORRECT_STATE;
+    default: // return CHIP_NO_ERROR and just read from the attribute store in default
+        break;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EnergyPrefAttrAccess::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    VerifyOrDie(aPath.mClusterId == EnergyPreference::Id);
+
+    // return CHIP_NO_ERROR and just write to the attribute store in default
+    return CHIP_NO_ERROR;
+}
+
+} // anonymous namespace
+
+void SetMatterEnergyPreferencesDelegate(EnergyPreferenceDelegate * aDelegate)
+{
+    gsDelegate = aDelegate;
+}
+
+EnergyPreferenceDelegate * GetMatterEnergyPreferencesDelegate()
+{
+    return gsDelegate;
+}
+
+Protocols::InteractionModel::Status
+MatterEnergyPreferenceClusterServerPreAttributeChangedCallback(const app::ConcreteAttributePath & attributePath,
+                                                               EmberAfAttributeType attributeType, uint16_t size, uint8_t * value)
+{
+    EndpointId endpoint                 = attributePath.mEndpointId;
+    EnergyPreferenceDelegate * delegate = GetMatterEnergyPreferencesDelegate();
+    uint32_t ourFeatureMap;
+    bool balanceSupported = (FeatureMap::Get(attributePath.mEndpointId, &ourFeatureMap) == EMBER_ZCL_STATUS_SUCCESS) &&
+        ((ourFeatureMap & to_underlying(Feature::kEnergyBalance)) != 0);
+    bool lowPowerSupported = (ourFeatureMap & to_underlying(Feature::kLowPowerModeSensitivity)) != 0;
+
+    if (delegate == nullptr)
+        return imcode::UnsupportedWrite;
+
+    switch (attributePath.mAttributeId)
+    {
+    case CurrentEnergyBalance::Id: {
+        if (balanceSupported == false)
+            return imcode::UnsupportedAttribute;
+
+        uint8_t index    = chip::Encoding::Get8(value);
+        size_t arraySize = delegate->GetNumEnergyBalances(endpoint);
+        if (index >= arraySize)
+        {
+            return imcode::InvalidValue;
```

Comment:
This is the deprecated name.  Please use ConstraintError, especially since that's what the spec says.